wire-ios-sync-engine/Tests/Source/Use cases/SetAllowGuestsAndServicesUseCaseTests.swift

```diff
@@ -0,0 +1,180 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import WireDataModelSupport
+import XCTest
+
+@testable import WireSyncEngine
+
+final class SetAllowGuestsAndServicesUseCaseTests: XCTestCase {
+
+    private let coreDataStackHelper = CoreDataStackHelper()
+    private var stack: CoreDataStack!
+    private let modelHelper = ModelHelper()
+    private var mockConversation: ZMConversation!
+    private var mockSelfUser: ZMUser!
+    private var sut: SetAllowGuestAndServicesUseCaseProtocol!
+
+    private var syncContext: NSManagedObjectContext {
+        return stack.syncContext
+    }
+
+    override func setUp() async throws {
+        try await super.setUp()
+        stack = try await coreDataStackHelper.createStack()
+        await syncContext.perform { [self] in
+            sut = SetAllowGuestAndServicesUseCase()
+            mockSelfUser = modelHelper.createSelfUser(in: syncContext)
+            mockConversation = modelHelper.createGroupConversation(in: syncContext)
+            mockConversation.teamRemoteIdentifier = UUID()
+        }
+    }
+
+    override func tearDown() async throws {
+        stack = nil
+        sut = nil
+        try coreDataStackHelper.cleanupDirectory()
+        try await super.tearDown()
+    }
+
+    func testGuestEnablementSucceeds() async {
+
+        await syncContext.perform { [self] in
+            // GIVEN
+            let role = Role.insertNewObject(in: syncContext)
+            let action = Action.insertNewObject(in: syncContext)
+            role.name = "wire_admin"
+            action.name = "modify_conversation_access"
+            role.actions = [action]
+
+            mockConversation.addParticipantAndUpdateConversationState(user: mockSelfUser, role: role)
+
+            let mockHandler = MockActionHandler<SetAllowGuestsAndServicesAction>(result: .success(()), context: syncContext.notificationContext)
+
+            let expectation = XCTestExpectation(description: "completion should be called")
+
+            // WHEN
+            sut.invoke(conversation: mockConversation, allowGuests: true, allowServices: false) { result in
+                // THEN
+                switch result {
+                case .success:
+                    print("Operation successful")
+                case .failure(let error):
+                    XCTFail("Test failed with error: \(error)")
+                }
+                expectation.fulfill()
+            }
+
+            wait(for: [expectation], timeout: 4.0)
+        }
+    }
+
+    func testGuestEnablementFails_WithInsufficientPermissions() async {
+        await syncContext.perform { [self] in
+            // GIVEN
+            let role = Role.insertNewObject(in: syncContext)
+            let action = Action.insertNewObject(in: syncContext)
+
+            action.name = "modify_conversation_access"
+            role.actions = [action]
```

Comment:
The only difference I see with the test where the user has permission is the role name. Is that gives the user permission?

Reply:
Yes, as far as I understood.

Reply:
Clean up the entire test class. I think now it's much more clear: https://github.com/wireapp/wire-ios/commit/80a04a5d4aaf473f97eee2a7d49763ac74f136c2